File: frontend/developer-workstation/src/utils/pkGenerationConfig.ts
export type PkGenerationStrategy = 'manual' | 'uuid' | 'autoIncrement' | 'prefixedSequence'

export interface PkGenerationConfig {
  strategy?: PkGenerationStrategy
  startValue?: number
  padWidth?: number
  prefix?: string
}

export const PK_GENERATION_STRATEGIES: PkGenerationStrategy[] = [
  'uuid',
  'manual',
  'autoIncrement',
  'prefixedSequence',
]

export function parsePkGeneration(
  raw?: Record<string, unknown> | PkGenerationConfig | null,
): PkGenerationConfig {
  if (!raw || typeof raw !== 'object') {
    return { strategy: 'uuid' }
  }
  const strategy = raw.strategy as PkGenerationStrategy | undefined
  return {
    strategy: PK_GENERATION_STRATEGIES.includes(strategy as PkGenerationStrategy)
      ? strategy
      : 'uuid',
    startValue: typeof raw.startValue === 'number' ? raw.startValue : 1,
    padWidth: typeof raw.padWidth === 'number' ? raw.padWidth : 6,
    prefix: typeof raw.prefix === 'string' ? raw.prefix : '',
  }
}

export function serializePkGeneration(
  config?: PkGenerationConfig | Record<string, unknown> | null,
  isPrimaryKey?: boolean,
): Record<string, unknown> | undefined {
  if (!isPrimaryKey) return undefined
  const parsed = parsePkGeneration(config)
  if (parsed.strategy === 'manual') {
    return { strategy: 'manual' }
  }
  if (parsed.strategy === 'uuid') {
    return { strategy: 'uuid' }
  }
  const out: Record<string, unknown> = {
    strategy: parsed.strategy,
    scope: 'perTable',
    startValue: parsed.startValue ?? 1,
  }
  if (parsed.strategy === 'prefixedSequence') {
    out.padWidth = parsed.padWidth ?? 6
    out.prefix = parsed.prefix ?? ''
  }
  return out
}

export function pkGenerationNeedsExtraConfig(strategy?: PkGenerationStrategy): boolean {
  return strategy === 'autoIncrement' || strategy === 'prefixedSequence'
}

File: backend/platform-common/src/main/java/com/platform/common/dto/PkGenerationConfig.java
package com.platform.common.dto;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;

/**
 * PK generation strategy stored on field metadata (PRD §5.2).
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class PkGenerationConfig {

    /** manual | uuid | autoIncrement | prefixedSequence | dailyDateSequence | monthlyDateSequence | customFormat */
    private String strategy;

    /** perTable | perFunctionUnit | perPrefix | perDay | perMonth */
    private String scope;

    private Long startValue;

    private Integer padWidth;

    private String prefix;

    /** Legacy {@code datePrefixedSequence} date style; migrated to {@code format} at allocate time. */
    private String datePattern;

    /** none | day | month — sequence reset for {@code customFormat}. */
    private String resetPeriod;

    /** Custom template for {@code customFormat}, e.g. {@code {DATETIME:yyyy-dd-MM}-{SEQNUM:4}}. */
    private String format;
}
